_posts/2021-09-03-rudiments-of-data-wrangling-in-rust.md

```diff
@@ -0,0 +1,519 @@
+
+The most challenging part of learning Rust is its memory management,
+which is based on tracking ownership and strict control over memory
+access, but doesn't have garbage collection like Java or Go. The Rust
+compiler has a feature called "the borrow checker", which is responsible
+for this control, so that Rust programs are as safe as possible at 
+runtime.
+
+Newcomers to Rust often find themselves initially fighting 
+the borrow checker, but once they get familiar with the rules, that
+changes into being grateful for yet another runtime disaster averted!
+
+The three phases of dealing with the Rust borrow checker, based on 
+the author's personal experience:
+
+1. Why the &%$! can't I do that?!
+2. Oh, I kind of see why that would be a problem.
+3. Thank you, borrow checker, for looking after my code!
+
+## Embrace the crates
+
+Every Rust program and library is packaged into a 
+[crate](https://doc.rust-lang.org/book/ch07-01-packages-and-crates.html), 
+which can refer to and depend on other crates. While Rust
+has a sizable [standard library](https://doc.rust-lang.org/std/), many interesting tasks
+can be achieved when you start taking advantage of the work
+that others have done.
+
+The [Cargo](https://doc.rust-lang.org/cargo/) tool is an essential part of the Rust tooling,
+and it is also responsible for resolving crate dependencies.
+Once you indicate in your program's `Cargo.toml` file that
+you would like to include a crate, Cargo picks it up,
+downloads it from the [crates.io](https://crates.io) registry, and compiles it 
+into your program.
+
+Many of the tasks in the small Rust program you are about to see
+are actually dependent on functionality found in crates
+such as the following:
+
+* [csv](https://crates.io/crates/csv) -- reading and writing CSV files
+* [encoding_rs](https://crates.io/crates/encoding_rs) -- implementation of the Encoding Standard
+* [encoding_rs_io](https://crates.io/crates/encoding_rs_io) -- transcoding from non-UTF-8 encoded content
+* [chrono](https://crates.io/crates/chrono) -- manipulating dates and times
+
+Some tasks like opening files, handling command-line arguments,
+and using data structures, remain the responsilibity of
+the Rust standard library. For more extensive handling of
+command-line arguments, take a look at [`clap`](https://github.com/clap-rs/clap) or [`structopt`](https://docs.rs/structopt/0.3.23/structopt/).
+
+Now, let's set the stage for the data wrangling.
+
+## Counting electric vehicle registrations
+
+Electric vehicles are rapidly becoming more common, and the pace will
+only pick up as we try to meet ambitious climate preservation goals
+also in Finland. I thought it would be interesting to know exactly 
+how quickly EVs have become more popular here in the recent years.
```

Comment:
Maybe here one could introduce the Finnish context a bit more? I'm thinking something like "my home country Finland", something to quickly explain to non Finnish readers why we're interested in Finland in particular, as opposed to, say, Europe as a whole.

Reply:
Good point, but decided to not push the home country aspect after all.